base/linalg/diagonal.jl

```diff
@@ -161,9 +161,14 @@ end
 # identity matrices via eye(Diagonal{type},n)
 eye{T}(::Type{Diagonal{T}}, n::Int) = Diagonal(ones(T,n))
 
-expm(D::Diagonal) = Diagonal(exp(D.diag))
-logm(D::Diagonal) = Diagonal(log(D.diag))
-sqrtm(D::Diagonal) = Diagonal(sqrt(D.diag))
+# Matrix functions
+^(D::Diagonal, p::Integer) = Diagonal((D.diag).^p)
```

Comment:
`^(D::Diagonal, p::Integer)` method is not necessary since its code is identical to `p::Real` case.